feat: Allow SelectViewType types field resolution

rafaeelaudibert opened the pull request:
Right now, we cannot resolve the proper type of view fields since we're only resolving them for `BaseTableType`. This is very restrictive and not very useful since we store type information for all saved views - and Revenue Analytics managed views - in the database and, therefore, have easy and accurate type information.

This is easy to see in production with the Revenue Analytics views, where we have some JSON fields - such as `metadata`. For JSON fields in a real "table" like the event `properties`, we can easily access them as `properties.key` and both the SQL editor and the actual query properly resolve that. If we do the same for `metadata.field`, then it fails because it does not know that `metadata` is a JSON field and doesn't apply the proper `extractRawJSON*` queries.

Here's a video explaning the whole situation
https://www.loom.com/share/510951d6f94e41159ef750bd6178dc0f?sid=dc858d15-b017-42f8-a412-a68aefce7f96

rafaeelaudibert committed:
feat: Allow `SelectViewType` types field resolution

Right now we cannot resolve the proper type of view fields since we're only resolving it for `BaseTableType`. This is very restrictive and not very useful since we store type information for all saved views - and Revenue Analytics managed views - in the database and, therefore, have easy and accurate type information.

This is easy to see in production with the Revenue Analytics views where we have some JSON fields - such as `metadata`. For JSON fields in a real "table" like the event `properties` we can easily access it as `properties.key` and both the SQL editor and the actual query properly resolve that. If we do the same for `metadata.field` then it fails because it does not know that `metadata` is a JSON field and doesn't apply the proper `extractRawJSON*` queries.

rafaeelaudibert committed:
feat: Add some tests to check metadata for views works

rafaeelaudibert commented on posthog/hogql/ast.py:
```diff
@@ -399,6 +407,18 @@ def print_type(self) -> str:
         return "String"
 
 
+@dataclass(kw_only=True)
+class StringJSONType(StringType):
+    def print_type(self) -> str:
+        return "JSON"
+
+
+@dataclass(kw_only=True)
+class StringArrayType(StringType):
+    def print_type(self) -> str:
+        return "Array"
```
This is something extra, happy to move to a separate PR and/or not include this. This makes the "type" on notices be more accurate

<img width="282" height="93" alt="image" src="https://github.com/user-attachments/assets/2aa9fc32-81f2-4f23-838e-7a740e72d83c" />

rafaeelaudibert committed:
refactor: Remove unneeded types

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert committed:
fix: Add missing `visit_decimal_type` visitor

rafaeelaudibert committed:
fix: Properly set fields for views

Since we now properly resolve view fields these were failing

Gilbert09 commented on posthog/hogql/ast.py:
```diff
@@ -506,7 +514,7 @@ class FieldType(Type):
 
     def resolve_database_field(self, context: HogQLContext) -> Optional[FieldOrTable]:
-        if isinstance(self.table_type, BaseTableType):
+        if isinstance(self.table_type, BaseTableType) or isinstance(self.table_type, SelectViewType):
```
Unsure on this case - a `SelectViewType` is not a `Table` in hogql land (hence it not being covered by `BaseTableType`)

rafaeelaudibert replied:
Yeah, I understand. We can solve this in a different way, but my problem is the following, this errors out when I'm trying to add a pretty join to a view

<img width="733" height="507" alt="image" src="https://github.com/user-attachments/assets/124535bd-4af2-4089-a893-3906769d34e9" />

It does work in the end, but it's in red

rafaeelaudibert committed:
tests: Expose error on debug mode

Makes testing simpler locally, just run DEBUG=1 pytest ... and you'll see the AST error

rafaeelaudibert committed:
fix: Add these back and update metadata test

These were removed because we did 2 separate PRs, add them here and update the test to reflect the JSON type

rafaeelaudibert committed:
refactor: Turn `SelectViewType` into a `BaseTableType`

These are always backed by a table so let's make that happen here. This will also guarantee we can resolve fields for views in the UI

rafaeelaudibert committed:
feat: Suggest now-supported `metadata.*` accessor

rafaeelaudibert committed:
fix: fix mypy error

rafaeelaudibert committed:
chore: Update mypy-baseline

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert committed:
fix: Clarify order when doing select star

rafaeelaudibert committed:
fix: Make sure workflow saved query includes columns
